lib/actions/auth.ts

```diff
@@ -0,0 +1,209 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { CommandDefinition } from 'capitano';
+
+export const login: CommandDefinition<
+	{},
+	{
+		token: string | boolean;
+		web: boolean;
+		credentials: boolean;
+		email: string;
+		password: string;
+	}
+> = {
+	signature: 'login',
+	description: 'login to balena',
+	help: `\
+Use this command to login to your balena account.
+
+This command will prompt you to login using the following login types:
+
+- Web authorization: open your web browser and prompt you to authorize the CLI
+from the dashboard.
+
+- Credentials: using email/password and 2FA.
+
+- Token: using a session token or API key from the preferences page.
+
+Examples:
+
+	$ balena login
+	$ balena login --web
+	$ balena login --token "..."
+	$ balena login --credentials
+	$ balena login --credentials --email johndoe@gmail.com --password secret\
+`,
+	options: [
+		{
+			signature: 'token',
+			description: 'session token or API key',
+			parameter: 'token',
+			alias: 't',
+		},
+		{
+			signature: 'web',
+			description: 'web-based login',
+			boolean: true,
+			alias: 'w',
+		},
+		{
+			signature: 'credentials',
+			description: 'credential-based login',
+			boolean: true,
+			alias: 'c',
+		},
+		{
+			signature: 'email',
+			parameter: 'email',
+			description: 'email',
+			alias: ['e', 'u'],
+		},
+		{
+			signature: 'password',
+			parameter: 'password',
+			description: 'password',
+			alias: 'p',
+		},
+	],
+	primary: true,
+	async action(_params, options, done) {
+		type Options = typeof options;
+		const _ = await import('lodash');
+		const Promise = await import('bluebird');
+		const balena = (await import('balena-sdk')).fromSharedOptions();
+		const auth = await import('../auth');
+		const form = await import('resin-cli-form');
+		const patterns = await import('../utils/patterns');
+		const messages = await import('../utils/messages');
+
+		const doLogin = (loginOptions: Options): Promise<void> => {
+			if (loginOptions.token != null) {
+				return Promise.try(() => {
+					if (_.isString(loginOptions.token)) {
+						return loginOptions.token;
+					}
+					return form.ask({
+						message: 'Session token or API key from the preferences page',
+						name: 'token',
+						type: 'input',
+					});
+				})
+					.then(balena.auth.loginWithToken)
+					.tap(() =>
+						balena.auth.whoami().then(username => {
+							if (!username) {
+								return patterns.exitWithExpectedError(
+									'Token authentication failed',
+								);
+							}
+						}),
+					);
+			} else if (loginOptions.credentials) {
+				return patterns.authenticate(loginOptions);
+			} else if (loginOptions.web) {
+				console.info('Connecting to the web dashboard');
+				return auth.login().return();
```

Comment:
The original code was `return auth.login()` -- I assume that `.return()` was needed to satisfy type checking. Nothing wrong, but given that this line relies on `login()` returning a Bluebird promise, I suggest either:
* Explicitly add a Bluebird return type to the `login()` function in `lib/auth/index.ts`, so at least it is clearer that it returns a Bluebird promise (instead of `async/await`), or
* Alternatively, replacing this line with: (I think)
```
    await auth.login();
    return;
```

Reply:
I'll look into also making use of async/await to simplify this

Reply:
done